x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ssl/SSLConfigurationReloaderTests.java

```diff
@@ -576,10 +593,14 @@ private void validateSSLConfigurationIsReloaded(
         // Baseline checks
         preChecks.accept(sslService.sslContextHolder(config).sslContext());
 
-        assertEquals("nothing should have called reload", 1, reloadLatch.getCount());
+        assertEquals("nothing should have called reload", modifierFunctions.size(), reloadLatch.getCount());
 
         // modify
-        modificationFunction.run();
+        for (var modifierFunction : modifierFunctions) {
+            modifierFunction.run();
+            Thread.sleep(RESOURCE_WATCHER_RELOAD_INTERVAL.millis() + 1);
```

Comment:
I confuses me slightly that this uses both a thread sleep and a count down latch.
I think we can write the test using only one of them, and prefer the latter.

Here's what I think should work:
```
-        final CountDownLatch reloadLatch = new CountDownLatch(modifierFunctions.size());
+        final CyclicBarrier reloadBarrier = new CyclicBarrier(2);
         final SSLService sslService = new SSLService(env);
         final SslConfiguration config = sslService.getSSLConfiguration("xpack.security.transport.ssl");
         final Consumer<SslConfiguration> reloadConsumer = sslConfiguration -> {
             try {
                 sslService.reloadSSLContext(sslConfiguration);
             } finally {
-                reloadLatch.countDown();
+                try {
+                    reloadBarrier.await();
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
             }
         };
         new SSLConfigurationReloader(reloadConsumer, resourceWatcherService, SSLService.getSSLConfigurations(env).values());
         // Baseline checks
         preChecks.accept(sslService.sslContextHolder(config).sslContext());

-        assertEquals("nothing should have called reload", modifierFunctions.size(), reloadLatch.getCount());
+        assertEquals("nothing should have called reload", 0, reloadBarrier.getNumberWaiting());

         // modify
         for (var modifierFunction : modifierFunctions) {
             modifierFunction.run();
-            Thread.sleep(RESOURCE_WATCHER_RELOAD_INTERVAL.millis() + 1);
+            reloadBarrier.await();
+            reloadBarrier.reset();
         }

-        reloadLatch.await();
         // checks after reload
         postChecks.accept(sslService.sslContextHolder(config).sslContext());
```
Alternatively we could have a list of count down latches, or, less preferred, a long sleep after all modifications (and no latches).

Reply:
Thanks for taking the time to improve on it! I agree, it would be ideal to drop the `sleep` so I'll change it to your solution. 

>a long sleep after all modifications

Yes, agreed, that's the most fragile but easiest to understand maybe. Anyways, I like that the barrier approach is so robust so let's go with that.